dev/core#241 Fix Manage Premiums UI

mattwire opened the pull request:
Overview
----------------------------------------
Manage premiums UI was not displaying the financial type correctly (it was showing the ID).  The PR does the following:
1. Enable pseudoconstant for financial_type_id.
2. Use pseudoconstant functions to get the financial_type label (name) for display.
3. Format the money fields using crmMoney.
4. Add the cost field to the UI (normally I'd do this as a separate PR but it would be more work to break out the template changes and given how flaky/old the premiums code seems to be I suspect it's not actually being used by very many people - also it's only affecting an admin screen by displaying existing information).

Before
----------------------------------------
- Financial type not displayed properly.
- Money not formatted properly.
![localhost_8000_civicrm_admin_contribute_managepremiums_reset 1 1](https://user-images.githubusercontent.com/2052161/42401277-d2445a5c-816c-11e8-8d5c-4563409272bb.png)


After
----------------------------------------
- Financial type not displayed properly.
- Money not formatted properly.
- Cost field displayed.
![localhost_8000_civicrm_admin_contribute_managepremiums_reset 1](https://user-images.githubusercontent.com/2052161/42401280-d61b90fa-816c-11e8-97b9-7a7a3ffcd438.png)

Financial type not reset to default everytime edit product on contribution page - added link to manually reset to default if required:
![localhost_8000_civicrm_admin_contribute_addproducttopage_action update id 1 pid 3 reset 1](https://user-images.githubusercontent.com/2052161/42523780-aca2ec4e-8466-11e8-8f5a-b53e15971e84.png)

Technical Details
----------------------------------------
The financial_type_id pseudoconstant can be used elsewhere in the code instead of relying on `CRM_Core_DAO::getFieldValue()`.

Comments
----------------------------------------
@pradpnayak Could you review this please?

mattwire commented:
test this please

pradpnayak commented on CRM/Contribute/Page/Premium.php:
```diff
@@ -166,9 +166,9 @@ public function browse() {
-          $premiums[$productDAO->id]['financial_type_id'] = CRM_Core_DAO::getFieldValue('CRM_Financial_DAO_FinancialType', $dao->financial_type_id, 'name');
+        // Financial Type
+        if (!empty($premiums[$productDAO->id]['financial_type_id'])) {
+          $premiums[$productDAO->id]['financial_type'] = CRM_Core_PseudoConstant::getLabel('CRM_Financial_BAO_FinancialType', 'financial_type', $premiums[$productDAO->id]['financial_type_id']);
```
When you add premiums to Contribution page you are allowed to change the Financial type (which is preselected from civicrm_product.financial_type_id). So this should be retrieved from civicrm_premiums_product not from civicrm_product.

pradpnayak commented:
retest this please

pradpnayak commented:
I did some test on current state of code. 

**Test case 1: Added New Product** 
Before:
![before1](https://user-images.githubusercontent.com/2053075/42459222-379f5a14-8393-11e8-9f4d-0c93f5c0b0a6.png)

After: Works fine
![after1](https://user-images.githubusercontent.com/2053075/42459369-6cdad726-8393-11e8-9603-5ad4e0c8070d.png)

**Test case 1: Added  Product Contribution Page** 
Before:
![before2](https://user-images.githubusercontent.com/2053075/42459463-8ca57e4e-8393-11e8-8507-c445be1bb0e1.png)

After: Financial Type are incorrect
![after2](https://user-images.githubusercontent.com/2053075/42459500-9f21b880-8393-11e8-9b01-de60265a1376.png)

pradpnayak commented:
Also found a bug when i edit a product from contribution page. The financial type is always defaulted from civicrm_product.financial_type_id on edit form rather from civicrm_premium_product.financial_type_id.

Before: FT should be Member dues
![before3](https://user-images.githubusercontent.com/2053075/42459751-3d966baa-8394-11e8-9aad-9486a948d94c.png)

After: FT should be Member dues
![after3](https://user-images.githubusercontent.com/2053075/42459756-40fb29b6-8394-11e8-85a5-32a2384aaa44.png)

mattwire commented:
@pradpnayak Thanks for review

> Also found a bug when i edit a product from contribution page. The financial type is always defaulted from civicrm_product.financial_type_id on edit form rather from civicrm_premium_product.financial_type_id.

I've fixed this (see second image in PR description AFTER).  There was some javascript that was resetting it after page load - now there's a link to reset it (using the same javascript) but it's not done automatically!

I've fixed the issue with reading financial type from the wrong place (as you said, should be premium_product, not product).  It confused me because there were two `$dao` objects in that function doing different things, I've now renamed them for clarity.

Could you review again please and confirm if this PR is ok now?

pradpnayak commented on templates/CRM/Contribute/Form/ContributionPage/AddProduct.tpl:
```diff
@@ -94,11 +94,10 @@
-    getFinancialType();
 
-    $("#product_id").change(getFinancialType);
+    $("#resetfinancialtype").click(getFinancialType);
```
We should return FALSE to prevent the default action.

pradpnayak commented on templates/CRM/Contribute/Form/ContributionPage/AddProduct.tpl:
```diff
@@ -94,11 +94,10 @@
         }
       });
-
+      $("#product_id").change(getFinancialType);
```
Won't this loop for infinity times since this called in getFinancialType()?

pradpnayak commented:
Hi Matt,

The changes look good. I did one more round of test on current state:
**Test case 1: Added New Product** 
Before:
![before1](https://user-images.githubusercontent.com/2053075/42459222-379f5a14-8393-11e8-9f4d-0c93f5c0b0a6.png)

After: Works fine
![a1](https://user-images.githubusercontent.com/2053075/42561721-becace88-84f1-11e8-80ac-58c72c09cfc8.png)


**Test case 2: Added  Product Contribution Page** 
Before:
![before2](https://user-images.githubusercontent.com/2053075/42459463-8ca57e4e-8393-11e8-8507-c445be1bb0e1.png)

After: Financial Type are correct
![a2](https://user-images.githubusercontent.com/2053075/42561732-c4648dac-84f1-11e8-9990-0354430f506e.png)

**Test case 3: Edit  Product from Contribution Page** 
Before: FT should be Member dues
![before3](https://user-images.githubusercontent.com/2053075/42459751-3d966baa-8394-11e8-9aad-9486a948d94c.png)

After: Financial Type is defaulted correctly(but still some issues)
![a3](https://user-images.githubusercontent.com/2053075/42561765-dc08fe98-84f1-11e8-8469-04b0e3c5ab55.png)


Still Problem:
1. When clicked on 'Reset to default'  the page scroll up to top.
2. When Product is changed on Edit screen Financial type is not changed initially need to click 'Reset to default' and then it is changed when product is changed.
3. After #2 there was approximately four calls to ajax to retrieve financial type for a selected product(should be one).

mattwire commented:
> Still Problem:
1. When clicked on 'Reset to default' the page scroll up to top.
1. When Product is changed on Edit screen Financial type is not changed initially need to click 'Reset to default' and then it is changed when product is changed.
1. After #2 there was approximately four calls to ajax to retrieve financial type for a selected product(should be one).

I've updated the javascript and template so that:
1. The reset "link" is only shown if the financial type is different.
2. The reset "link" does not cause the page to scroll use (name= instead of href=).
@pradpnayak What do you think?
3. The AJAX call is only called once.

pradpnayak commented:
Thanks @mattwire for the fix. I did the QA and it works fine now.
![1](https://user-images.githubusercontent.com/2053075/42637627-98107822-85e3-11e8-8b1b-309fe4d3ee59.png)

1. The reset link only appears when FT ic changed
![2](https://user-images.githubusercontent.com/2053075/42637643-a0517270-85e3-11e8-9609-5256a3cbdd38.png)

2. When clicked on reset link the page is not loaded

3. AJAX call is one per click.

Over to @eileenmcnaughton for final review and merge approval.

eileenmcnaughton commented:
I took a look at the code and nothing raises any red flags to me. @pradpnayak testing appears thorough - merging
